keylime/tpm/tpm_main.py

```diff
@@ -913,7 +914,11 @@ def tpm_init(self, self_activate: bool = False, config_pw: Optional[str] = None)
         # if no AIK created, then create one
         self.__create_aik()
 
-        return self.get_tpm_metadata("ekcert"), self.get_tpm_metadata("ek_tpm"), self.get_tpm_metadata("aik_tpm")
+        return (
+            cast(Optional[bytes], self.get_tpm_metadata("ekcert")),
+            cast(Optional[bytes], self.get_tpm_metadata("ek_tpm")),
+            cast(Optional[bytes], self.get_tpm_metadata("aik_tpm")),
```

Comment:
every cast is a red flag, I am sure that here too

Reply:
I am not sure, the `get_tpm_metadata` return the value of a YAML, and AFAIK there is no `bytes` representation in YAML (there are arrays, dates, string, integers ...). So for sure what ever `get_tpm_metadata` it is returning it is not a `bytes`.

What I am falling to communicate is that this cast is not doing any conversion at all. You are pbly returning an object that is a string, but you are telling to the type checker that consider it as a `Optional[bytes]` for the type check about the return value.

If you read this carefully, this would help:

```python
>>> from typing import cast, Any
>>> a:Any = "1"
>>> __annotations__
{'a': typing.Any}
>>> b:int = cast(int, a)
>>> __annotations__
{'a': typing.Any, 'b': <class 'int'>}
>>> b
'1'
```

`b` is type `int` but we are storing a `str`. That is what is wrong with `cast`

Reply:
We do get `bytes`, This here for example sets the "ekcert":

**https://github.com/keylime/keylime/blob/64e44d859608b69e03c970f30ff9841afe66752d/keylime/tpm/tpm_main.py#L910-L911**

https://github.com/keylime/keylime/blob/64e44d859608b69e03c970f30ff9841afe66752d/keylime/tpm/tpm_main.py#L1451

Reply:
> We do get bytes, This here for example sets the "ekcert":

Exactly, and this is why the `cast` is plain wrong.

Please, bear with me here : )))

Why are we doing all this type annotation work? Because we want an external tool to probe that the constrains imposed by the types are kept, so the invariant that they attach can be assumed. Basically here means that "ekcert" will be `None` or an array of bytes, so I can do for example a `base64.b64encode` without any conversion .

So the idea is that once variable and function signatures are annotated, the tool can follow the flow a make those proofs for us.

For the "ekcert" case we have the local variable `ekcert` with the correct type and the `tpm_init` return signature with the correct annotation. But the proof fails because the function `get_tpm_metadata` has a too generic return type, as has the `dict` internally represented.

Using a `cast` here invalidate the issues that pyright / mypy are pointing wrt to the `{set,get}_tpm_metadata`, the YAML document, or the internal `dict`. For example, I can make a mistake during a refactor and do in line 912 a `ekcert = "mytestcert"`, and `pyright` will not detect the change of type from bytes to string, hiding the exact type of bugs that we want to show.

For the specific `ekcert` case a better solution should be to return `ekcert` directly, without calling the metadata function, so we can connect the type of the variable with the return value, and drop one single cast. A better solution is to fix the metadata API and see if it can be fixed, or transform the `dict` into a `dataclass` with types.

The problem here is that each `cast` is an indication of a different bug that we are hiding, so will require a per-case fix.

AFAIK the tpm_* classes would be deeply refactored, so in this case (and in others) IMHO we should not go too far with the annotations, and wait until the refactor.

Also as a general idea, if there is no other alternative than a `cast`, an `assert` should be there too. But every `cast` is a red flag.

Reply:
I 'fixed' this now without much refactoring, which should happen in a separate patch. For now I added asserts.

Reply:
When we remove the agent code (hopefully this month), most of the code in tpm_* will be also removed.

After the removal we should discuss on how to refactor the TPM validation and event loop code because both require massive changes to implement for example the push model idea or if we want to support other forms of attestation.